### ISSUE TITLE
CoreFX's version of CLI should be updated automatically when a new LKG of CLI is available

### ISSUE THREAD

**Issue by dagood** [dotnet/corefx#7952]
For more info on the plan, skip to this comment: https://github.com/dotnet/corefx/issues/7952#issuecomment-213072085

This issue was originally about the following bug that we're fairly sure was because we had an out of date CLI:

---

Error during restore in [this OSX Release build](http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/osx_release_prtest/28/):

```
Tools are already initialized
  [14:24:51.24] Restoring all packages...
  /tmp/tmp62f36ecd1afa49f8a8396d696921c22e.exec.cmd: line 3:  4667 Segmentation fault: 11  "/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx_release_prtest/Tools/dotnetcli/dotnet" restore --packages "/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx_release_prtest/packages" --source https://dotnet.myget.org/F/dotnet-core/api/v3/index.json --source https://www.nuget.org/api/v2/ --infer-runtimes "/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx_release_prtest/src" "/Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx_release_prtest/pkg"
  [14:28:23.61] Restoring all packages...Done.
```

So far this has only been observed in that one linked build, so I don't know if it's a one-off issue or not.

Summary of ideas on this (thanks @ericstj @mellinoe)

This was found in https://github.com/dotnet/corefx/pull/7931. However, https://github.com/dotnet/corefx/issues/7902 indicated that there was a corrupted package in the NuGet cache, which suggests that some sort of interruption earlier caused a package to be corrupted during restore. It may have been this same kind of segfault during a package install that caused the corrupt package, but we don't have a good way of tracking this down.

A simple `./build.sh managed release` doesn't repro the segfault on my OSX box.

One thing to try is to update CLI to take any fixes to the runtime that might have happened.

(Side effect: NuGet may also want to be resilient to corrupted cache files, in case segfaults or even just build cancellations happen when the cache is being written.)

**Comment by stephentoub** [comment]
@dagood, did the version of CLI being used include a CoreCLR with one of these fixes?
https://github.com/dotnet/coreclr/pull/4341 (rc2)
https://github.com/dotnet/coreclr/pull/4266 (rc3)

**Comment by ericstj** [comment]
The CLI being used is from end of march: https://github.com/dotnet/corefx/blob/master/DotnetCLIVersion.txt

**Comment by dagood** [comment]
This is the tree for `1.0.0-beta-002173`: https://github.com/dotnet/cli/tree/b3e8612812be11b193c49f542cd27657af027ac6 but I don't see where the CoreCLR version is. It's definitely older than 7 days though!

**Comment by ericstj** [comment]
Yeah.

**Comment by stephentoub** [comment]
My guess is that's the problem.  This issue was causing seg faults multiple times a day on OSX in some our of more intensive test suites (e.g. PLINQ, System.Linq.Expressions), and since the fix went in, I don't think we've seen any seg faults in CI runs (or at least I haven't noticed any).

What's our plan for updating to more recent CLIs and staying up-to-date so that we're always on the latest stable version?

**Comment by dagood** [comment]
For keeping our CLI version up to date, if it gets put on https://github.com/eerhardt/versions we could have those PRs automated just like the dependency upgrades. /cc @eerhardt

**Comment by weshaggard** [comment]
>What's our plan for updating to more recent CLIs and staying up-to-date so that we're always on the latest stable version?

We don't have a strategy for this yet. We are essentially on the plan of updating on demand when needed. I'm not convinced yet that we want to be picking up a version automatically, but it is something we can consider for our dev branch to help with dogfooding CLI.

**Comment by ericstj** [comment]
I think it should be automatic via PR.

**Comment by eerhardt** [comment]
Can I update this issue title to "Automatically update CoreFX with a new CLI"?  Or is this issue supposed to be to fix the OSX problem right now?

**Comment by eerhardt** [comment]
But on the topic of automatic updating - this will fit in nicely with the other automatic PRs we are doing.  We will just need to:

0. Create the `dotnet/versions` repo
1. Write out the CLI build version to the `dotnet/versions` repo when new builds are available
2. Update our https://github.com/dotnet/corefx/blob/master/UpdateDependencies.ps1 script to pull in the new CLI build number as well.  (The 2nd half of the script will take care of creating the PR.)
3. Add a subscription to https://github.com/eerhardt/versions/blob/master/Subscriptions.json when a new CLI build is out to kick off the CoreFX "update dependencies" build.

**Comment by dagood** [comment]
I just put in a PR for the CLI upgrade, so I'm fine with this issue being reused for the new infrastructure. And that plan sounds good, but for (3), it seems like we'll also need to pass something in (or separate out the scripts) to distinguish between CLI version and CoreFX version changes. Maybe that's what you mean, but if we don't do that it sounds like CLI upgrades would be tied to package upgrades which I don't think we want.

**Comment by eerhardt** [comment]
I think we could do it either way, it doesn't necessarily have to be decided right now.

The plan @piotrpMSFT and I had for keeping the CLI repo up-to-date with its dependencies was to check all downstream dependencies every time.  And for all of them that have changed, upgrade.

The advantage this gives is:

1. As soon as one comes out, you will start a PR with the new dependency version.
2. If before #1 finishes another new dependency updates, you kick off a new PR with both dependencies updated, and you can cancel #1.
 - This gets you to a place with all of your dependencies updated and verified ASAP. As opposed waiting for both PRs to go through, and then doing another run with them both together.

CoreFX may have different needs/goals, and we can do whatever is best for CoreFX here.  Just enumerating this strategy, to give you some thoughts.

**Comment by dagood** [comment]
For a successful flow that works, I'm just thinking about if/when CLI includes breaking changes that we haven't fully prepared for. (Removing `--infer-runtimes` or compile-only TxMs comes to mind.) At that point I'm thinking we wouldn't want to be blocked on upgrading CLI to get new corefx dependency versions. That's an exceptional case though, and even with no automatic support for it we could always run dependency upgrades manually until all the auto-upgrades work. So on second thought, that plan looks good to me.

**Comment by eerhardt** [comment]
Assigning to @dagood, as I won't be making these changes.

**Comment by karelz** [comment]
@mellinoe can you please move it to buildtools? This is impacting more than just corefx repo.

**Comment by mellinoe** [comment]
@dagood Is there a better place to track this?

**Comment by mellinoe** [comment]
We discussed this briefly today and we don't really think there's much to be done here. We don't think we want to automatically be updating the CLI version, but we should still be thinking about periodically doing so in the future, especially (only?) after major versions are released.

**Comment by dagood** [comment]
Even if we only update it periodically and intentionally, the PRs could still be sent out automatically. Then if there are errors for some product repos, the generated branches can be used as the basis for fixing up the repo. Doing updates less often just changes the desired update trigger to "stable" versions rather than "LKG", where the current "stable" version might just be updated manually.

I'll create a new issue for this (in a private repo at the moment because that's where we're tracking stuff).

**Comment by karelz** [comment]
We should be very careful about automatic PRs cadence - we want to avoid noise from them.
